src/components/Modal/BaseModal.js

```diff
@@ -24,173 +26,188 @@ const defaultProps = {
-        // To handle closing any modal already visible when this modal is mounted, i.e. PopoverReportActionContextMenu
-        Modal.setCloseModal(this.props.onClose);
-    }
-
-    componentDidUpdate(prevProps) {
-        if (prevProps.isVisible === this.props.isVisible) {
-            return;
-        }
-
-        Modal.willAlertModalBecomeVisible(this.props.isVisible);
-        Modal.setCloseModal(this.props.isVisible ? this.props.onClose : null);
-    }
-
-    componentWillUnmount() {
-        // Only trigger onClose and setModalVisibility if the modal is unmounting while visible.
-        if (this.props.isVisible) {
-            this.hideModal(true);
-            Modal.willAlertModalBecomeVisible(false);
-        }
-
-        // To prevent closing any modal already unmounted when this modal still remains as visible state
-        Modal.setCloseModal(null);
-    }
+function BaseModal({
+    isVisible,
+    onClose,
+    shouldSetModalVisibility,
+    onModalHide,
+    type,
+    popoverAnchorPosition,
+    innerContainerStyle,
+    outerStyle,
+    onModalShow,
+    propagateSwipe,
+    fullscreen,
+    animationIn,
+    animationOut,
+    useNativeDriver,
+    hideModalContentWhileAnimating,
+    animationInTiming,
+    animationOutTiming,
+    statusBarTranslucent,
+    onLayout,
+    avoidKeyboard,
+    forwardedRef,
+    children,
+}) {
+    const {windowWidth, windowHeight, isSmallScreenWidth} = useWindowDimensions();
+
+    const safeAreaInsets = useSafeAreaInsets();
 
     /**
      * Hides modal
      * @param {Boolean} [callHideCallback=true] Should we call the onModalHide callback
      */
-    hideModal(callHideCallback = true) {
-        if (this.props.shouldSetModalVisibility) {
-            Modal.setModalVisibility(false);
-        }
-        if (callHideCallback) {
-            this.props.onModalHide();
+    const hideModal = useCallback(
+        (callHideCallback = true) => {
+            if (shouldSetModalVisibility) {
+                Modal.setModalVisibility(false);
+            }
+            if (callHideCallback) {
+                onModalHide();
+            }
+            Modal.onModalDidClose();
+            if (!fullscreen) {
+                ComposerFocusManager.setReadyToFocus();
+            }
+        },
+        // eslint-disable-next-line react-hooks/exhaustive-deps -- adding onModalHide to the dependency array causes too many unnecessary rerenders
+        [shouldSetModalVisibility],
+    );
+
+    useEffect(() => {
+        Modal.willAlertModalBecomeVisible(isVisible);
+
+        // To handle closing any modal already visible when this modal is mounted, i.e. PopoverReportActionContextMenu
+        Modal.setCloseModal(isVisible ? onClose : null);
+
+        return () => {
+            // Only trigger onClose and setModalVisibility if the modal is unmounting while visible.
+            if (isVisible) {
+                hideModal(true);
+                Modal.willAlertModalBecomeVisible(false);
+            }
```

Comment:
This _will_ cause a regression. We should not use the isVisible state here as this will be evaluated at the time of render and not when the component unmounts i.e. if you initially render the modal as visible (true) then on unmount this condition will always be true even if the state value change by then.

What we should do instead is to use a ref

```js
    const isVisibleRef = useRef(isVisible);
    useEffect(() => {
        isVisibleRef.current = isVisible;
    }, [isVisible]);
```